### DISCUSSION TITLE
Syntax for strings

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#17] [General]
In the case of [RDF 1.1 Turtle](https://www.w3.org/TR/2014/REC-turtle-20140225/#sec-grammar-grammar) specification,
```
[17] String            ::= STRING_LITERAL_QUOTE | STRING_LITERAL_SINGLE_QUOTE | STRING_LITERAL_LONG_SINGLE_QUOTE | STRING_LITERAL_LONG_QUOTE
[18] IRIREF                                ::= '<' ([^#x00-#x20<>"{}|^`\]     | UCHAR)* '>' /* #x00=NULL #01-#x1F=control codes #x20=space */
[22] STRING_LITERAL_QUOTE                  ::= '"' ([^#x22#x5C#xA#xD] | ECHAR | UCHAR)* '"' /* #x22=" #x5C=\ #xA=new line #xD=carriage return */
[23] STRING_LITERAL_SINGLE_QUOTE           ::= "'" ([^#x27#x5C#xA#xD] | ECHAR | UCHAR)* "'" /* #x27=' #x5C=\ #xA=new line #xD=carriage return */
[24] STRING_LITERAL_LONG_SINGLE_QUOTE ::= "'''" (("'" | "''")? ([^'\] | ECHAR | UCHAR))* "'''"
[25] STRING_LITERAL_LONG_QUOTE        ::= '"""' (('"' | '""')? ([^"\] | ECHAR | UCHAR))* '"""'
[26] UCHAR                            ::= '\u' HEX HEX HEX HEX | '\U' HEX HEX HEX HEX HEX HEX HEX HEX
[159s] ECHAR                          ::= '\' [tbnrf"'\]
[171s] HEX                            ::= [0-9] | [A-F] | [a-f]
```

Similarly, in the [SPARQL 1.1](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#sparqlGrammar) grammar:
```
[135] String         ::= STRING_LITERAL1 | STRING_LITERAL2 | STRING_LITERAL_LONG1 | STRING_LITERAL_LONG2
[139] IRIREF                 ::= '<' ([^<>"{}|^`\]-[#x00-#x20])* '>'
[156] STRING_LITERAL1        ::= "'" ( ([^#x27#x5C#xA#xD]) | ECHAR )* "'"
[157] STRING_LITERAL2        ::= '"' ( ([^#x22#x5C#xA#xD]) | ECHAR )* '"'
[158] STRING_LITERAL_LONG1 ::= "'''" ( ( "'" | "''" )? ( [^'\] | ECHAR ) )* "'''"
[159] STRING_LITERAL_LONG2 ::= '"""' ( ( '"' | '""' )? ( [^"\] | ECHAR ) )* '"""'
[160] ECHAR                ::= '\' [tbnrf\"']
```

Our syntax is inspired by them.

([SPARQL 1.2](https://www.w3.org/TR/sparql12-query/#grammar) Working Draft is also available, but it's the same in this regard.)

**Comment by nichtich** [comment]
Difference to pure JSON string syntax:

- single quoted strings: in my opinion this has no real benefit but opens room for endless fights about which one to prefer. Furthermore single quote is allowed in URIs so this would collide with unquoted identifiers
- Extended Unicode escape sequence with `\U'
- multiline-strings

**Comment by nichtich** [comment]
CYPHER uses `  for strings, including multiline-strings, this could be supported in PG format the same way, but I'm not sure whether it's actually needed:

~~~
a key: `this values spans
multiple lines and an
escape `` character.`
~~~

~~~json
{"id":"a","labels":[],"properties":{"key":["this values spans\nmultiple lines and an\nescaped ` character."]}}
~~~

**Comment by hchiba1** [comment]
When dealing with large text data in the real world, one comes across strings that include various combinations of quotation marks.
Simple converters of such text data into PG format may struggle with escaping these quotation marks properly.

Triple quotes would be useful as a last resort for dealing with such messy strings
(because even real-world strings rarely contain both `"""` and `'''` simultaneously).

**Reply by nichtich** [reply]
In real world data is always messy and contains errors and special cases in many suprising ways, so I would not make any assumptions :grin:

In any case I would stick to existing syntax (JSON, Cypher, and/or Turtle). I've chosen JSON because the PG data model is serialized in JSON anyway, so implementations need to know about JSON strings anyway. Cypher also seems reasonable because people working with PG will likely also know or learn about Cypher, and parsing/encoding rule is on line of code each. If you choose Turtle syntax then exactely as it is although this requires more complex implementation (but doable). We could use examples from the [W3C Turtle Test suite](https://www.w3.org/2013/TurtleTests/) to check.

**Comment by nichtich** [comment]
Without multiline strings, a parser can tokenize input line-by-line. Line breaks can still be done with `"\n"`. On the other hand multiline strings are more readable for humans so they should be supported in some way. As far as I understand the way multiline strings are defined in Turtle and SPARQL with `"""` and `'''`, it is not possible to properly encode the string `"\"\"\"\n'''"` as multiline string, is it?. Using ``` ` ```. as in Cypher just requires  to replace every ``` ` ```  with ``` `` ```, this seems easier and it also comes from property graph community.

**Reply by hchiba1** [reply]
I agree with the need for multiline strings.

By the way, where are the specs for the double backticks (``` `` ```)?
(Is it a common way to escape `` ` ``? Maybe specific to Cypher.)

**Reply by nichtich** [reply]
See https://opencypher.org/resources/ in particular https://s3.amazonaws.com/artifacts.opencypher.org/M23/railroad/SymbolicName.html#EscapedSymbolicName indiectly referenced from https://s3.amazonaws.com/artifacts.opencypher.org/M23/railroad/Atom.html

**Comment by hchiba1** [comment]
Is it a bad idea to allow all control characters (including newline `#xA`) in quoted strings
so that the multiline strings are allowed?

See the updated https://github.com/pg-format/pg-formatter/blob/main/docs/grammar.md
where almost all characters are allowed in the quoted strings (except the same quotes as themselves).
Invalid Unicode sequences might be included in the input data,
but I think the applications can inspect the content of parsed strings, raising errors if needed.

**Reply by nichtich** [reply]
Well, it's a matter of choice to allow newlines in quoted strings. I've rarely seen it in practice. One example is GraphViz Dot format but I only found out via a bug report where a software did not support newlines in quoted strings although they are not forbidden.

On the other hand, if it's documented well, we can remove special character `` ` `` and just use `"`. In this case we extend JSON string syntax where control characters U+0000 through U+001F must be escaped!

I'm less sure about `'` but for another reason. It is [reserved character](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2) in URI so it is allowed (in contrast to `` ` `` , `"` and other such as `|`, `^`, `<`, `>`) so it should not be part of `UNQUOTED_CHAR`. Disallowing `'` only as the first character of an unquoted string could be ok.

Without single quoted strings and `` ` `` this could be:

~~~ebnf
String         ::= QuotedString | UnquotedString
QuotedString   ::= '"' ( UNESCAPED | '\' ( ["\/bfnrt] | 'u' HEX HEX HEX HEX ) )* '"'
UnquotedString ::= UNQUOTED_FIRST UNQUOTED_CHAR*
UNQUOTED_FIRST ::= [^\x20\x09\x0A\x0D"<>{}|^:'(#-]
UNQUOTED_CHAR  ::= [^\x20\x09\x0A\x0D"<>{}|^]
UNESCAPED      ::= [^\x00-\x08\x0B\x0C\x0E-\x1F'\\]  # allow \t \n \r
HEX            ::= [0-9a-fA-F]
~~~

**Reply by hchiba1** [reply]
In the specs of RDF/SPARQL (excerpted at the beginning of this discussion page), 
control characters are excluded from IRIREF, but not excluded from strings.

And surprisingly, the definition of IRIREF is different between RDF and SPARQL: 
Unicode sequences are used in RDF, but not in SPARQL.

**Reply by nichtich** [reply]
Oh, standards can be a nightmare!

**Reply by nichtich** [reply]
Allowing control characters would make parsing more difficult but serializing easier:  just replace all `"` with `\"` and enclose the value in `"`.

**Comment by hchiba1** [comment]
I would like to support `'`.
I think this style of quotation is too prevalent to be excluded.

**Reply by nichtich** [reply]
Well, `'` it allowed for instance in CYPHER but not in DOT or JSON - it's a pure matter of choice.

Would you support control characters (that means multiline-strings) in both `"` and `'` quoted strings or only when using `"`?

**Reply by hchiba1** [reply]
In the end, I would like to propose `'` `"` disallowing multiline-strings and `'''` `"""` allowing multiline-strings.

**Reply by nichtich** [reply]
HiWhen `"`" quoted string disallow linebreaks, I'd prefer to keep JSON compatibility with its escaping rules such as `\"`, `\n` `\uXXXX`. But what are the rules of `'` quoted strings? If we allow both `\"` and `\'` in single and double quoted strings, we leave JSON territory, as `"\'"` is no valid JSON string! Nevertheless I've updated pgraph, so single quoted strings are supported since version 0.5.3.

I looked at existing JSON supersets such as [JSON5](https://spec.json5.org/#strings) and found [CoffeeScript-Object-Notation (CSON)](https://github.com/bevry/cson?tab=readme-ov-file#what-is-cson) seems very similar to what you propose:

- double quoted strings and single quoted strings
- escape sequences as JSON strings but extended:
   - `\` is also allowed to escape other character, e.g. `\x` and `\'` (I'd only allow `\'`)
   - for some reason `\n` where `n` is a number is not allowed (I'd not copy this)
   - `\` followed by spaces and newline is ignored (I'd not copy this)
- multiline-strings with `"""` and `'''` can include same escape sequences
- longes common sequence of space before each line in multiline-strings is ignore
- space-only first and last line is ignored in multiline-strings

Example:

~~~
a
  key:"""
    example of
      \"""-escaped
    string
  """
~~~

would be

~~~
a key:'example of\n  """-escaped\nstring'
~~~

but `` ` `` instead of `"""` and `'''` and no escaping would also be ok - this looks cleaner to me:

~~~
a
  key:`
    example of
      """-escaped
    string
  `
~~~

**Comment by nichtich** [comment]
Coming back to this issue with some distance.

Fixed:

- `"` and `'` can be used to quote strings
- JSON escape sequences in strings quoted with `"` and `'`

Open

- Whether quoted strings with `"` and `'` can span multiline lines
- Whether another method is used for multiline strings and whether escape sequences are allowed these strings

One use case of multiline strings not mentioned before is to allow copy & pase into syntax. This would require to *not* allow escape sequences but act like here doc document in shell or like [code spans in Markdown](https://spec.commonmark.org/0.31.2/#code-spans). 

At the moment I lean to either allow line breaks in `"` and `'` quoted strings and not introduce another string syntax, or disallow linebreaks in these strings and a new "verbatim string" type that requires no escaping at all.

**Reply by hchiba1** [reply]
I think it is OK to use `"` and `'` for multi-line strings too.

Such multi-line string representations seem to be observed in some languages like PHP and Ruby.

**Reply by nichtich** [reply]
So strings can use escape rules (e.g. `\t`, `\r`, `\n`) but also directly include these control characters. The following two strings would all be equivalent:

~~~
"\n\n"

'\n\n'

"
\n"

'\n
'
~~~

**Comment by nichtich** [comment]
Verbatim string syntax adjusted from CommonMark specification:

- A **backtick sequence** is a string of one or more backtick characters (`` ` ``) that is neither preceded nor followed by a backtick.

- A **verbatim string** begins with a backtick sequence and ends with a backtick sequence of equal length. The string value of the verbatim string is the characters between these two backtick sequences. If the string both begins and ends with a line break, both line breaks are removed from the string value.

Examples:

~~~
# "foo"
`foo`
# "foo ` bar"  (line breaks stripped)
``
foo ` bar
``
# "foo ` bar\n"  (line break not stripped)
``foo ` bar
``
# "foo\\nbar"
`foo\nbar`
~~~

**Reply by hchiba1** [reply]
Such spec seems to be a bit complicated to me.

**Reply by nichtich** [reply]
Well that's how Markdown has been specified and I think it's a very clever method to allow arbitrary strings without escaping. I [implemented it yesterday](https://github.com/pg-format/pgraphs/commit/972e975b17843c57649764af0e76626395997ee0#diff-1e24cba3952977b598c0c7f9623cc5cb3f28a163be5e2a0f9de45c783f0cad08R144-R157) without problem, but sure could roll back the feature.

**Comment by hchiba1** [comment]
How to evaluate the escape sequences in `'` is a matter of choice. 

How about differentiating the use of escape sequences in `'`?
That is, escape sequences are not evaluated in `'`.
This is a traditional style in shell script, Perl, and similar languages.

Examples:
```
# 'a\nb'
'a\nb'

# "a\nb"
"a
b"
```

**Reply by nichtich** [reply]
That's one string syntax for quoted strings (`"` with escape sequences)  and another syntax for verbatim strings. It only remains to decide on rules for verbatim strings. I don't mind whether to use `'` or `` ` ``.

**Comment by nichtich** [comment]
I've adjusted both pgraphs, pg-formatter, and [the specification draft section on quoted strings](https://pg-format.github.io/specification/#quoted-strings) to use the same definition: JSON strings but also `'` can be used (=single quoted) and newlines, carriage return and tabulator are allowed (=multiline strings). The use of `` ` `` is not included, this could be added as extension later.

By this I'd like to close the discussion for now, to get to PG format 1.0.0.
